lib/operations/add_user.js

```diff
@@ -1,186 +1,171 @@
-      if (err && err.code !== -5000) return handleCallback(callback, err);
 
-      // We need to perform the backward compatible insert operation
-      const finalOptions = applyWriteConcern(Object.assign({}, options), { db }, options);
+    // Check the db name and add roles if needed
+    if (
+      (db.databaseName.toLowerCase() === 'admin' || finalOptions.dbName === 'admin') &&
+      !Array.isArray(finalOptions.roles)
+    ) {
+      roles = ['root'];
+    } else if (!Array.isArray(finalOptions.roles)) {
+      roles = ['dbOwner'];
+    }
+
+    // Get the error options
+    const commandOptions = {};
+
+    const writeConcern = WriteConcern.fromOptions(finalOptions);
+    if (writeConcern != null) {
+      commandOptions.writeConcern = writeConcern;
+    }
+
+    if (finalOptions.dbName) {
+      commandOptions.dbName = finalOptions.dbName;
+    }
+
+    // Add maxTimeMS to options if set
+    if (typeof finalOptions.maxTimeMS === 'number') {
+      commandOptions.maxTimeMS = finalOptions.maxTimeMS;
+    }
+
+    const digestPassword = db.s.topology.lastIsMaster().maxWireVersion >= 7;
+
+    let userPassword = password;
 
+    if (!digestPassword) {
       // Use node md5 generator
       const md5 = crypto.createHash('md5');
       // Generate keys used for authentication
       md5.update(username + ':mongo:' + password);
-      const userPassword = md5.digest('hex');
+      userPassword = md5.digest('hex');
+    }
 
-      // If we have another db set
-      const dbToUse = options.dbName ? new Db(options.dbName, db.s.topology, db.s.options) : db;
+    // Build the command to execute
+    const command = {
+      createUser: username,
+      customData: finalOptions.customData || {},
+      roles: roles,
+      digestPassword
+    };
+
+    // No password
+    if (typeof password === 'string') {
+      command.pwd = userPassword;
+    }
 
-      // Fetch a user collection
-      const collection = dbToUse.collection(CONSTANTS.SYSTEM_USER_COLLECTION);
+    // Error out if digestPassword set
+    if (finalOptions.digestPassword != null) {
+      return callback(
+        toError(
+          "The digestPassword option is not supported via add_user. Please use db.command('createUser', ...) instead for this option."
+        )
+      );
+    }
 
-      // Check if we are inserting the first user
-      count(collection, {}, finalOptions, (err, count) => {
-        // We got an error (f.ex not authorized)
-        if (err != null) return handleCallback(callback, err, null);
-        // Check if the user exists and update i
-        const findOptions = Object.assign({ projection: { dbName: 1 } }, finalOptions);
-        collection.find({ user: username }, findOptions).toArray(err => {
+    // Attempt to execute auth command
+    super.execute(command, (err, r) => {
+      if (!err) {
+        return handleCallback(callback, err, r);
+      }
+
+      if (err && err.ok === 0 && err.code === undefined) {
```

Comment:
since we're here, and we don't support 2.4 or lower, can we just get rid of this?

Reply:
is all of this just for 2.4 support?

Reply:
Yeah, everything inside the `if (err && err.ok === 0 && err.code === undefined) {` is to catch the case where the server doesn't know the `createUser` command.